### ISSUE TITLE
When the number of Kafka messages is too large, the TraceId is confused.

### ISSUE THREAD

**Issue by HuangDayu** [apache/skywalking#5781]
Please answer these questions before submitting your issue.

- Why do you submit this issue?

- [x] Bug

___
### Question
- What do you want to know?

**zh_cn**: 我想要跟踪一帧TCP消息的生产消费过程，但是当消息量过大时，Kafka的消费者的TraceId则絮乱了，出现了大量重复。

**en**: I want to track the production and consumption process of a frame of TCP messages, but when the message volume is too large, the TraceId of Kafka consumers is confused, and there are a lot of repetitions.

___
### Bug
- Which version of SkyWalking, OS and JRE?

|     name     |     version     |
| :----------: | :-------------: |
|      os      |     `win10`     |
|     java     |   `openjdk11`   |
|  SkyWalking  |     `8.2.0`     |
|    kafka     |     `2.5.0`     |
| spring-cloud |  `Hoxton.SR8`   |
| spring-boot  | `2.3.2.RELEASE` |
| spring-kafka | `2.5.4.RELEASE` |

- What happened?

#### tcp threads is 100 ，100 messages

**producer**
![image](https://user-images.githubusercontent.com/23421574/97965406-4662b880-1df5-11eb-9621-95409f60d647.png)

**consumer**

![image](https://user-images.githubusercontent.com/23421574/97965609-9e012400-1df5-11eb-9d3f-5104b5ac156e.png)

#### tcp threads is 1000 ，1000 messages

**producer**

![image](https://user-images.githubusercontent.com/23421574/97965702-bf621000-1df5-11eb-893a-d01caf1d0bf3.png)

**consumer**

![image](https://user-images.githubusercontent.com/23421574/97965733-cb4dd200-1df5-11eb-96ce-aee66781dde2.png)

___
### Requirement or improvement
- Please describe your requirements or improvement suggestions.

I am not familiar with implementation logic.

**Comment by wu-sheng** [comment]
What do you TCP message and Kafka? Tracing TCP frame is not a good design, but I can't link this to Kafka.

**Comment by HuangDayu** [comment]
为了跟踪IOT设备指令的生产与消费，消息内容如：`helloABCDEFDCBA`，kafka是集群模式，2.5.0版本。

In order to track the production and consumption of IOT device instructions, the message content is such as: `helloABCDEFDCBA`, Kafka is a cluster mode, version 2.5.0

我重写了ChannelInboundHandlerAdapter的channelRead方法，并且添加了`@Trace`注解，得以跟踪TCP消息，代码如下。

I override the channelRead method of ChannelInboundHandlerAdapter and add the @Trace annotation to trace TCP messages. The code is as follows.

```java
    @Trace
    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) {
        if (msg instanceof String) {
            String command = (String) msg;
            log.info("channel read message:{}", command);
            ctx.writeAndFlush("\n");
            kafkaService.sendMessageToAdapter(new QueueMessageDTO(instanceId, "xxx", "xxx", "xxx", command));
        }
        ReferenceCountUtil.release(msg);
    }
```

**Comment by HuangDayu** [comment]
- success

![image](https://user-images.githubusercontent.com/23421574/97969263-c4758e00-1dfa-11eb-941a-43dab5d32894.png)

- exception

![image](https://user-images.githubusercontent.com/23421574/97969432-0272b200-1dfb-11eb-920f-dc2e12ef6aba.png)

**Comment by wu-sheng** [comment]
If you have changed the codes, it is hard to tell why trace-id duplicated. 
____
Here are my guesses and suggestions only.
First, are these segment/messages belong to one context you do? Because you add `@Trace`, so maybe that is your own code issue, which is putting the root span in the wrong place.
In the Kafka consumer side, there would be no root span, unless you set one like `@Trace`.
Also, you need to find, whether the multiple messages really carry the same context. 
Above are my best guesses.

Basically, if you changed the codes, and are using the codes in some scenarios we don't know, I recommend we work on some high-level discussion, rather than codes and UI. I am not familiar with IoT data collection, if you could provide a graph or more context about what is your workflow and data flow. We may discuss based on that.
Instead, we should discuss, ho

**Comment by HuangDayu** [comment]
消费者和生产者在同一个环境下，也就是 WIN10，kafka集群在linux环境下。

Consumers and producers are in the same environment, which is WIN10, and the Kafka cluster is in the Linux environment.

经过测试，在相同代码下，TCP线程数620以上是，则开始出现TID重复的情况。消息体并没有出现重复，这似乎有一定规律。

After testing, under the same code, if the number of TCP threads is 620 or more, the TID duplicates begin to appear. There is no repetition in the message body, which seems to have a certain pattern.

**Comment by wu-sheng** [comment]
Don't check the TID in the log only, that thing could be affected by many uncertain things. Check the `sw8` in the message header, if those are duplicated, then you are facing some issues at the client-side.

**Comment by HuangDayu** [comment]
How check the sw8 in the message header？

**Comment by wu-sheng** [comment]
That is Kafka, you could read directly. We are using the Kafka message header to propagate the context, including the trace id.

**Comment by HuangDayu** [comment]
What is header's key?

**Comment by HuangDayu** [comment]
# Producer

## code

```java
	@Trace
    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) {
        if (msg instanceof String) {
            String command = (String) msg;
            ctx.writeAndFlush("\n");
            String uuid = UUID.randomUUID().toString();
            String traceId = TraceContext.traceId();
            TraceContext.putCorrelation("swTraceId", traceId);
            TraceContext.putCorrelation("uuid", uuid);
            log.info("swTraceId:{},uuid:{}, message:{},targetId:{}", traceId, uuid, command, uuid);
            kafkaService.sendMessageToAdapter(new QueueMessageDTO(instanceId, uuid, "one", "one", command));
        }
        ReferenceCountUtil.release(msg);
    }
```

## log

```log
2020-11-03 19:35:22,767  INFO 38616 TID:58cfe19c15954d1999e5e30e48f4e327.211.16044033227670349 [ntLoopGroup-5-8] c.n.i.c.g.t.a.h.InboundMessageHandler    : swTraceId:58cfe19c15954d1999e5e30e48f4e327.211.16044033227670349,uuid:98a2dc07-90bf-47c4-bf0d-05cf7ffde412, message:helloABCDEFDCBA,targetId:98a2dc07-90bf-47c4-bf0d-05cf7ffde412
```

# Consumer

## code

```java
    @Trace
    @KafkaListener(topics = "${spring.application.name}", groupId = "${spring.application.name}")
    private void receiveMessage(ConsumerRecord<?, ?> record) {
        Optional<?> kafkaMessage = Optional.ofNullable(record.value());
        if (kafkaMessage.isPresent()) {
            Object message = kafkaMessage.get();
            QueueMessageDTO queueMessageDTO = JSONUtil.toBean(String.valueOf(message), QueueMessageDTO.class);
            log.info("swTraceId:{},swTraceId:{},uuid: {} ,message: {},targetId:{}", TraceContext.traceId(), TraceContext.getCorrelation("swTraceId"),
                    TraceContext.getCorrelation("uuid"), queueMessageDTO.getCommand(), queueMessageDTO.getTargetServiceId());
        }
    }
```

## log

```log
2020-11-03 19:35:22,770  INFO 36440 TID:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347 [ntainer#0-0-C-1] c.n.i.c.p.a.app.service.KafkaService     : swTraceId:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347,swTraceId:Optional[58cfe19c15954d1999e5e30e48f4e327.211.16044033227670349],uuid: Optional[98a2dc07-90bf-47c4-bf0d-05cf7ffde412] ,message: helloABCDEFDCBA,targetId:1e8672b0-4b4e-47a3-a06c-0a5a4d2c695a
2020-11-03 19:35:22,771  INFO 36440 TID:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347 [ntainer#0-0-C-1] c.n.i.c.p.a.app.service.KafkaService     : swTraceId:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347,swTraceId:Optional[58cfe19c15954d1999e5e30e48f4e327.211.16044033227670349],uuid: Optional[98a2dc07-90bf-47c4-bf0d-05cf7ffde412] ,message: helloABCDEFDCBA,targetId:4501b5d2-d0ef-4321-889a-2e6be0bfb6ad
2020-11-03 19:35:22,771  INFO 36440 TID:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347 [ntainer#0-0-C-1] c.n.i.c.p.a.app.service.KafkaService     : swTraceId:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347,swTraceId:Optional[58cfe19c15954d1999e5e30e48f4e327.211.16044033227670349],uuid: Optional[98a2dc07-90bf-47c4-bf0d-05cf7ffde412] ,message: helloABCDEFDCBA,targetId:a9949f6e-2f6f-4fd5-a274-38ec17f37ef5
2020-11-03 19:35:22,771  INFO 36440 TID:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347 [ntainer#0-0-C-1] c.n.i.c.p.a.app.service.KafkaService     : swTraceId:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347,swTraceId:Optional[58cfe19c15954d1999e5e30e48f4e327.211.16044033227670349],uuid: Optional[98a2dc07-90bf-47c4-bf0d-05cf7ffde412] ,message: helloABCDEFDCBA,targetId:988d86fa-5148-4021-b483-70e27c77d820
2020-11-03 19:35:22,771  INFO 36440 TID:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347 [ntainer#0-0-C-1] c.n.i.c.p.a.app.service.KafkaService     : swTraceId:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347,swTraceId:Optional[58cfe19c15954d1999e5e30e48f4e327.211.16044033227670349],uuid: Optional[98a2dc07-90bf-47c4-bf0d-05cf7ffde412] ,message: helloABCDEFDCBA,targetId:c365fb85-23f5-4706-9afe-4c60490b9451
2020-11-03 19:35:22,771  INFO 36440 TID:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347 [ntainer#0-0-C-1] c.n.i.c.p.a.app.service.KafkaService     : swTraceId:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347,swTraceId:Optional[58cfe19c15954d1999e5e30e48f4e327.211.16044033227670349],uuid: Optional[98a2dc07-90bf-47c4-bf0d-05cf7ffde412] ,message: helloABCDEFDCBA,targetId:395d469b-d174-43c8-9c52-5c7918d2bf09
2020-11-03 19:35:22,771  INFO 36440 TID:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347 [ntainer#0-0-C-1] c.n.i.c.p.a.app.service.KafkaService     : swTraceId:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347,swTraceId:Optional[58cfe19c15954d1999e5e30e48f4e327.211.16044033227670349],uuid: Optional[98a2dc07-90bf-47c4-bf0d-05cf7ffde412] ,message: helloABCDEFDCBA,targetId:1c85e97d-6797-46ed-a7be-76e731f070f8
2020-11-03 19:35:22,771  INFO 36440 TID:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347 [ntainer#0-0-C-1] c.n.i.c.p.a.app.service.KafkaService     : swTraceId:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347,swTraceId:Optional[58cfe19c15954d1999e5e30e48f4e327.211.16044033227670349],uuid: Optional[98a2dc07-90bf-47c4-bf0d-05cf7ffde412] ,message: helloABCDEFDCBA,targetId:45109ff9-9178-4019-a777-e90aa32f4555
2020-11-03 19:35:22,771  INFO 36440 TID:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347 [ntainer#0-0-C-1] c.n.i.c.p.a.app.service.KafkaService     : swTraceId:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347,swTraceId:Optional[58cfe19c15954d1999e5e30e48f4e327.211.16044033227670349],uuid: Optional[98a2dc07-90bf-47c4-bf0d-05cf7ffde412] ,message: helloABCDEFDCBA,targetId:2d46da9e-4699-4710-a86b-cf2470c18708
2020-11-03 19:35:22,771  INFO 36440 TID:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347 [ntainer#0-0-C-1] c.n.i.c.p.a.app.service.KafkaService     : swTraceId:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347,swTraceId:Optional[58cfe19c15954d1999e5e30e48f4e327.211.16044033227670349],uuid: Optional[98a2dc07-90bf-47c4-bf0d-05cf7ffde412] ,message: helloABCDEFDCBA,targetId:47f433a9-5d10-49a0-aee3-b41e34326707
2020-11-03 19:35:22,774  INFO 36440 TID:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347 [ntainer#0-0-C-1] c.n.i.c.p.a.app.service.KafkaService     : swTraceId:58cfe19c15954d1999e5e30e48f4e327.209.16044033227560347,swTraceId:Optional[58cfe19c15954d1999e5e30e48f4e327.211.16044033227670349],uuid: Optional[98a2dc07-90bf-47c4-bf0d-05cf7ffde412] ,message: helloABCDEFDCBA,targetId:98a2dc07-90bf-47c4-bf0d-05cf7ffde412

```

# Test result

我修改代码做了一个测试，得出一下结论：

- 线程数是700，消费者和生产者得到的日志也是700行，且消息对象中的`targetId`并没有重复，这说明消息并没有重复；
- 通过`TraceContext.putCorrelation();`方法传递的`swTraceId`,`uuid`也是重复的；
- 在生产者日志中，`swTraceId`与`traceId`是一致的，但是到消费者中却不一致了。

**也就是说，kafka传递的消息体正常，但是Trace上下文全乱了，包括自定义的Correlation。**

---

I modified the code and did a test and came to the following conclusion:

- The number of threads is 700, the logs obtained by consumers and producers are also 700 lines, and the `targetId` in the message object is not repeated, which means that the message is not repeated;The number of threads is 700, the logs obtained by consumers and producers are also 700 lines, and the `targetId` in the message object is not repeated, which means that the message is not repeated;

- The `swTraceId` and `uuid` passed by the `TraceContext.putCorrelation();` method are also repeated;

- In the producer log, `swTraceId` and `traceId` are the same, but in the consumer they are not consistent.

**In other words, the message body passed by Kafka is normal, but the Trace context is completely messed up, including the custom Correlation.**

**Comment by wu-sheng** [comment]
First, if the trace context in the header of Kafka's message is correct, then this is not a bug of SkyWalking and its plugin.
The issue you are facing because of your `@trace` and your codes. 

---
In the SkyWalking, if a segment(consumer in your case) has multiple parents(we call it `reference`), then only the first trace-id left to propagate. Because this is better for performance. If you are using the batch consuming mode, then this is expected, not a bug. If you are not expecting the batch consumption, then it means you put `@trace` at the wrong place.
---

I hope the above explanation is clear enough for you. Actually, this is tracing 101 :)

**Comment by HuangDayu** [comment]
意思是kafka主题监听方法中添加​@Trace注解会导致批处理？印象中即使我删除了该注解，依然会得到上面的结果。我明天重复认证一下。

**Comment by wu-sheng** [comment]
Please keep your comments in dual languages or English. The discussion should be readable for people in the world.

**Comment by HuangDayu** [comment]
进过测试，即使取消掉消费者端的`@trace`注解，一旦消息量过大时，依然出现这个问题。

In my tests, even if The '@trace' annotation on the consumer side was removed, this problem still occurred once the message volume was too large.

**Comment by wu-sheng** [comment]
If you understand how Kafka consumer works, and read this FAQ, https://github.com/apache/skywalking/blob/master/docs/en/FAQ/kafka-plugin.md. From my understanding, there is nothing SkyWalking could do to make this happens, unless this is the real case.

It really needs your capability to locate the issue. Kafka consumer-side only creates one segment with one span only, unless, you put a bigger scape segment/context to include.

**Comment by HuangDayu** [comment]
# Producer Logger

```log
2020-11-04 19:19:04,443  INFO 14208 TID:5464bdfc220741749ed2eb3d2157879d.175.16044887444430451 [ntLoopGroup-5-7] c.n.i.c.g.t.a.h.InboundMessageHandler    : traceId:5464bdfc220741749ed2eb3d2157879d.175.16044887444430451,uuid:496296fc-868c-4c82-97c4-95295c9c40af, message:helloABCDEFDCBA,targetId:496296fc-868c-4c82-97c4-95295c9c40af
```

# Consumer Logger

```log
2020-11-04 19:19:04,447  INFO 5944 TID:5464bdfc220741749ed2eb3d2157879d.173.16044887444390451 [ntainer#0-0-C-1] c.n.i.c.p.a.app.service.KafkaService     : kafka headers sw8 values:[, 5464bdfc220741749ed2eb3d2157879d.175.16044887444430451, 5464bdfc220741749ed2eb3d2157879d.924.16044887444440154, , spring-cloud-app, 0515084b1cd74aada0a5bcd95fed7988@192.168.2.168, SpringAsync, 192.168.2.223:9092;192.168.2.224:9092;192.168.2.226:9092]
2020-11-04 19:19:04,447  INFO 5944 TID:5464bdfc220741749ed2eb3d2157879d.173.16044887444390451 [ntainer#0-0-C-1] c.n.i.c.p.a.app.service.KafkaService     : traceId:5464bdfc220741749ed2eb3d2157879d.173.16044887444390451,uuid: 496296fc-868c-4c82-97c4-95295c9c40af ,message: helloABCDEFDCBA,targetId:496296fc-868c-4c82-97c4-95295c9c40af
```

我修改了测试代码，将`kafka`的`header`的`sw8`的值输出，得到了一个惊人的结果：`sw8`的值中的traceId是正确的，但是`TraceContext.traceId()`的值却是错误的，且是之前消息的traceId，`TraceContext.getCorrelation(key)`也是如此。可以看出，kafka传输的header是没有问题的，至于为什么`TraceContext`对象的值却是旧的值，我不知如何debug，目前所以找不到问题所在。

I have modified the test code, will ` kafka ` ` of header of ` ` sw8 ` value output, received a surprising result: ` sw8 ` traceId in value is correct, but ` TraceContext. TraceId () ` value is wrong, and is traceId before the news, the ` TraceContext. GetCorrelation ` (key).As you can see, kafka's header is ok. As for why the value of 'TraceContext' object is old, I don't know how to debug, so I can't find the problem at present.

@wu-sheng

**Comment by wu-sheng** [comment]
This is not a surprising result, this is what I told you. 
![image](https://user-images.githubusercontent.com/5441976/98110321-e09c2c80-1ed9-11eb-904b-916bf7c8272d.png)

**Comment by HuangDayu** [comment]
TraceId will not be enabled if I do not add the @trace annotation to the Kafka producer.

**Comment by wu-sheng** [comment]
Sorry, I can't do more for your private codes. I have given all information I know.

**Comment by HuangDayu** [comment]
也许是我不能够完全明白你的意思，不过，非常感谢你耐心的回复。祝好。

Maybe I can't understand you completely, but thank you very much for your patient reply.I wish good.

**Comment by akwangl** [comment]
哥们我跟你遇到同样的问题，kafka头是正常的

Buddy, I encountered the same problem as you. The kafka header is normal

> 也许是我不能够完全明白你的意思，不过，非常感谢你耐心的回复。祝好。
> 
> Maybe I can't understand you completely, but thank you very much for your patient reply.I wish good.
